Brooks/NewLanguageRenderer.hs

```diff
@@ -0,0 +1,675 @@
+
+plusEqualsDocD :: (Doc, Maybe String) -> (Doc, Maybe String) -> Doc
+plusEqualsDocD (v1, _) (v2, _) = v1 <+> text "+=" <+> v2
+
+plusEqualsDocD' :: (Doc, Maybe String) -> Doc -> (Doc, Maybe String) -> Doc
+plusEqualsDocD' (v1, _) plusOp (v2, _) = v1 <+> equals <+> v1 <+> plusOp <+> v2
+
+plusPlusDocD :: (Doc, Maybe String) -> Doc
+plusPlusDocD (v, _) = v <> text "++"
+
+plusPlusDocD' :: (Doc, Maybe String) -> Doc -> Doc
+plusPlusDocD' (v, _) plusOp = v <+> equals <+> v <+> plusOp <+> int 1
+
+varDecDocD :: Label -> Doc -> Doc
+varDecDocD l st = st <+> text l
+
+varDecDefDocD :: Label -> Doc -> (Doc, Maybe String) -> Doc
+varDecDefDocD l st (v, _) = st <+> text l <+> equals <+> v
+
+listDecDocD :: Label -> (Doc, Maybe String) -> Doc -> Doc
+listDecDocD l (n, _) st = st <+> text l <+> equals <+> new <+> st <> parens n
+
+listDecDefDocD :: Label -> Doc -> [(Doc, Maybe String)] -> Doc
+listDecDefDocD l st vs = st <+> text l <+> equals <+> new <+> st <+> braces (callFuncParamList vs)
+
+objDecDefDocD :: Label -> Doc -> (Doc, Maybe String) -> Doc
+objDecDefDocD = varDecDefDocD
+
+constDecDefDocD :: Label -> Doc -> (Doc, Maybe String) -> Doc -- can this be done without StateType (infer from value)?
+constDecDefDocD l st (v, _) = text "const" <+> st <+> text l <+> equals <+> v
+
+returnDocD :: (Doc, Maybe String) -> Doc
+returnDocD (v, _) = text "return" <+> v
+
+commentDocD :: Label -> Doc -> Doc
+commentDocD cmt cStart = cStart <+> text cmt
+
+freeDocD :: (Doc, Maybe String) -> Doc
+freeDocD (v, _) = text "delete" <+> v
+
+throwDocD :: Doc -> Doc
+throwDocD errMsg = text "throw new" <+> text "System.ApplicationException" <>
+    parens errMsg
+
+statementDocD :: (Doc, Terminator) -> (Doc, Terminator)
+statementDocD (s, t) = (s <> getTermDoc t, Empty)
+
+getTermDoc :: Terminator -> Doc
+getTermDoc Semi = semi
+getTermDoc Empty = empty
+
+-- Unary Operators --
+
+notOpDocD :: Doc
+notOpDocD = text "!"
+
+notOpDocD' :: Doc
+notOpDocD' = text "not"
+
+negateOpDocD :: Doc
+negateOpDocD = text "-"
+
+sqrtOpDocD :: Doc
+sqrtOpDocD = text "sqrt"
+
+sqrtOpDocD' :: Doc
+sqrtOpDocD' = text "math.sqrt"
+
+absOpDocD :: Doc
+absOpDocD = text "fabs"
+
+absOpDocD' :: Doc
+absOpDocD' = text "math.fabs"
+
+logOpDocD :: Doc
+logOpDocD = text "log"
+
+logOpDocD' :: Doc
+logOpDocD' = text "math.log"
+
+lnOpDocD :: Doc
+lnOpDocD = text "ln"
+
+lnOpDocD' :: Doc
+lnOpDocD' = text "math.ln"
+
+expOpDocD :: Doc
+expOpDocD = text "exp"
+
+expOpDocD' :: Doc
+expOpDocD' = text "math.exp"
+
+sinOpDocD :: Doc
+sinOpDocD = text "sin"
+
+sinOpDocD' :: Doc
+sinOpDocD' = text "math.sin"
+
+cosOpDocD :: Doc
+cosOpDocD = text "cos"
+
+cosOpDocD' :: Doc
+cosOpDocD' = text "math.cos"
+
+tanOpDocD :: Doc
+tanOpDocD = text "tan"
+
+tanOpDocD' :: Doc
+tanOpDocD' = text "math.tan"
+
+asinOpDocD :: Doc
+asinOpDocD = text "asin"
+
+asinOpDocD' :: Doc
+asinOpDocD' = text "math.asin"
+
+acosOpDocD :: Doc
+acosOpDocD = text "acos"
+
+acosOpDocD' :: Doc
+acosOpDocD' = text "math.acos"
```

Comment:
This seems like a Java-ism ? So it seems out of place.

Reply:
It's used in Python and Lua (though we haven't written the Lua renderer yet). Since more than one language uses these (in the long run), I think it's okay to keep them where they are?